src/SecureStore.HashicorpVault/HashicorpVaultSecureStore.cs

```diff
@@ -0,0 +1,357 @@
+                    "delete");
+            }
+            catch (SecureStoreException sse) when (sse.ErrorType == SecureStoreException.Type.SecretNotFound)
+            {
+                // Ignore SecretNotFound
+            }
+        }
+
+        public async Task<string> CreateValueAsync(string context, string key, string value)
+        {
+            var ctx = ConvertJsonToContext(context);
+
+            // key is null for new secret
+            value = value ?? throw new ArgumentNullException(nameof(value));
+
+            await ExecuteHashicorpVaultOperation(
+                async () =>
+                {
+                    var client = _clientFactory.CreateClient(ctx);
+                    return await client.SetSecretAsync(key, value);
+                },
+                "set");
+
+            return key;
+        }
+
+        public async Task<Credential> GetCredentialsAsync(string context, string key)
+        {
+            var ctx = ConvertJsonToContext(context);
+            key = key ?? throw new SecureStoreException();
+
+            var secret = await ExecuteHashicorpVaultOperation(
+                async () =>
+                {
+                    var client = _clientFactory.CreateClient(ctx);
+                    return await client.GetCredentialAsync(key);
+                },
+                "get");
+
+            return secret;
+        }
+
+        public async Task<string> CreateCredentialsAsync(string context, string key, Credential value)
+        {
+            var ctx = ConvertJsonToContext(context);
+            key = key ?? throw new ArgumentNullException(nameof(key));
+
+            await ExecuteHashicorpVaultOperation(
+                async () =>
+                {
+                    var client = _clientFactory.CreateClient(ctx);
+                    return await client.SetCredentialAsync(key, value);
+                },
+                "set");
+            return key;
+        }
+
+        public async Task<string> UpdateValueAsync(string context, string key, string oldAugmentedKey, string value)
+        {
+            var ctx = ConvertJsonToContext(context);
+            key = key ?? throw new ArgumentNullException(nameof(key));
+            oldAugmentedKey = oldAugmentedKey ?? throw new ArgumentNullException(nameof(oldAugmentedKey));
+            value = value ?? throw new ArgumentNullException(nameof(value));
+
+            if (key.Equals(oldAugmentedKey, StringComparison.Ordinal))
+            {
+                await ExecuteHashicorpVaultOperation(
+                    async () =>
+                    {
+                        var client = _clientFactory.CreateClient(ctx);
+                        return await client.SetSecretAsync(key, value);
+                    },
+                    "set");
+            }
+            else
+            {
+                await CreateValueAsync(context, key, value);
+                await RemoveValueAsync(context, oldAugmentedKey);
```

Comment:
Should we catch the exception and retry if `RemoveValueAsync` throws an exception when the new secret has already been created.